sacred/observers/s3_observer.py

```diff
@@ -0,0 +1,328 @@
+#!/usr/bin/env python
+# coding=utf-8
+
+import json
+import os
+import os.path
+
+import boto3
+from botocore.errorfactory import ClientError
+
+from sacred.commandline_options import CommandLineOption
+from sacred.dependencies import get_digest
+from sacred.observers.base import RunObserver
+from sacred.serializer import flatten
+import re
+import socket
+
+DEFAULT_S3_PRIORITY = 20
+
+
+def _is_valid_bucket(bucket_name):
+    # See https://docs.aws.amazon.com/awscloudtrail/latest/userguide/
+    # cloudtrail-s3-bucket-naming-requirements.html
+    if len(bucket_name) < 3 or len(bucket_name) > 63:
+        return False
+    if '..' in bucket_name or '.-' in bucket_name or '-.' in bucket_name:
+        return False
+    for char in bucket_name:
```

Comment:
I think this could be expressed more concisely using a regex. Also the  "cannot end with a dash" requirement could be expressed then.

Reply:
I've refactored this somewhat in the commits responding to comments; I do feel like the "allowed character" check is easier to read and parse in current form than it would be in a regex, which are generally a little opaque to understand. But if you still prefer a regex to the updated version, I can switch it.